src/main/java/com/vaadin/recipes/recipe/blockingdialog/BlockingDialog.java

```diff
@@ -0,0 +1,104 @@
+            dialog.setCloseOnOutsideClick(false); // allowing to remove the yes/no buttons would keep the user-code un-executed forever, leaking memory and possibly other resources.
+            dialog.open();
+
+            // Command will be run in the UI thread eventually
+            final YesNoDecider.MyCommand command = new YesNoDecider.MyCommand() {
+                @Override
+                public void execute() {
+                    if (future.isDone()) {
+                        try {
+                            consumer.accept(future.get());
+                            done = true;
+                        } catch (InterruptedException | ExecutionException ex) {
+                            ex.printStackTrace();
+                        }
+                    }
+                }
+            };
+
+            UI ui = UI.getCurrent(); // in the UI thread, get the UI instance
+            new Thread(() -> { // worker thread trying to let the UI thread run the command which in turn runs the user code as soon as the required user-input boolean is there
+                UI.setCurrent(ui); // in case the command uses 'UI.getCurrent()' assign the parent thread's UI to this new child thread.
+                while (!command.done) { // repeat until it is finally done.
+                    try {
+                        ui.access(command).get(); // and use the UI instance in a child thread, block this thread until the UI thread tried another time.
+                        // network usage: make sure that there is no polling set so that we only piggy-back upon the client-server communications that take
+                        // place after some user action such as a click.
+                        // Using Vaadin Push, we would not need such a piggyback: one could push UI changes out to the client using UI.push().
+                    } catch (InterruptedException | ExecutionException ex) {
+                        ex.printStackTrace();
+                    }
+                }
+            }).start(); // one may consider instead using a thread pool, see ExecutorService for example.
+
+            // functional identity mapping with a little side effect
+            future.thenApply(value -> {
+                dialog.close();
+                return value;
+            });
+        }
+    }
+
+
+    public BlockingDialog(){
+
+        // Consider the implicit Consumer<Boolean> below as the user code that can in principle be arbitrarily complex.
+        // In Swing, it would be all the application code following the user's decision -- the 'continuation'.
+        // The main work migrating a Swing application's blocking dialog is to refactor that continuation into the
+        // Consumer instance.
+        Button openDialog = new Button("Open a seemingly *blocking* dialog", event -> YesNoDecider.yesNo("Would you like to save the world?", answer -> {
```

Comment:
In what way is this blocking? If it would be blocking, then we wouldn't get to the `add(openDialog)` line until after the user has clicked either yes or no.

If it isn't blocking, then this is effectively just a confirm dialog which can be implemented with regular click handlers and no threading tricks.

Reply:
It *is* effectively just a confirm dialog. However, the point is that the running thread's code is cut so that it immediately ends; while another thread is picking up the dangling end after customer choice.

In a project that is designed in a reactive way, such as a native Vaadin programming style, this is implemented in a straight-forward fashion using the regular click handlers and no threading tricks, agreed.

I am trying here, though, to bridge the mental gap between "main application loop" thinking and the need to 'release' a running thread so it can finalize the HTTP response.

I wonder how I can make this clearer, after all I am simulating a problem to solve it -- so it's presence has to be assumed; if it's not, the solution is useless by definition.

Reply:
Oh, and it *is* blocking the application from being operated due to forcing one of the dialog buttons to be pushed, other dialog closing options disabled, dialog being modal and all.

Reply:
There are two aspects of dealing with a Swing-style genuinely blocking confirm dialog. The first thing is that the user cannot do anything else in the application. That's already covered by disabling alternative ways of dismissing the dialog.

The second this is that the the programming model looks like an `if` over the result rather than as two separate callback handlers. This cannot directly work with the underlying request/response nature unless you move _everything_ to a separate thread. What can be done is to still have a typical event handling callback that is run asynchronously, but promise to only run it once and pass a parameter to that callback that encourages an `if`. In this way, application code would follow this structure:
```
showConfirmDialog("Are you sure").then(confirmed -> {
  if (confirmed) {
    Notification.show("Good");
  } else {
    Notification.show("Don't be shy");
  }
});
```

Anything beyond that leads to a situation where `Notification.show` would run outside something that is a proper UI thread. This is possible, but it's quite a mess with synthetically moving lock ownership to the other thread and then moving it back to the request thread after the original thread completes. With that approach, application code could look like this:
```
magic(confirmationDialogController -> {
  if (confirmationDialogController.show("Are you sure")) {
    Notification.show("Good");
  } else {
    Notification.show("Don't be shy");
  }
});
```

Reply:
The way it's currently implemented (including the lock "simplification" pushed a couple of minutes ago) is in the first category when it comes to the structure of application code. It doesn't need a thread at all. All you need is that  the "Yes" button does `consumer.accept(true)` and the "No" button does `consumer.accept(false)`.

Reply:
> This cannot directly work with the underlying request/response nature unless you move everything to a separate thread

This is exactly what I was doing lately for a customer, but it becomes clear to me that now that it is done -- it is fine to not let a self-created slave thread carry that 'Command'-pattern style code capsule with the 'everything' continuation (very very far branching-out application code), but instead let some future UI thread deal with it as soon as the user made their decision.
As I mentioned in the documentation, the biggest work is to turn directly called code into a Consumer<Boolean>, or even Consumer<Something> to make it a call-back. e.g. for a button click handler.

Now that this code collapsed into a basic Dialog example with two buttons, there is no value in it as a cookbook recipe. I may instead write an article on refactoring techniques for the transformation needed.

I'll get in touch about the sense or absence thereof regarding the access() task queue, but this is no longer the right place.

Thanks for the review.